components/ingest-service/grpc/grpc.go

```diff
@@ -83,19 +93,29 @@ func Spawn(opts *serveropts.Opts) error {
 		return err
 	}
 
-	if migrationNeeded {
-		err = migrator.Start()
-		if err != nil {
-			log.WithError(err).Error("Failed starting a migration")
-			return err
+	migrationRunning, err := migrator.MigrationRunning()
+	if err != nil {
+		log.WithError(err).Error("Failed checking if a migration is running")
+		return err
+	}
+
+	if !migrationRunning {
+		if migrationNeeded {
+			err = migrator.Start()
```

Comment:
There could be a small gap of time between calling `migrator.MigrationRunning()` and calling `migrator.Start()`, and another front-end could claim the lock during that time. The documentation (https://www.postgresql.org/docs/9.6/functions-admin.html#FUNCTIONS-ADVISORY-LOCKS) is not as explicit as I would prefer but it says that `pg_advisory_lock` will block/wait until the lock becomes available. 

One thing you could do instead is to use `pg_try_advisory_lock()` which returns a boolean (false if you didn't get the lock). Then your program flow here might be like

```go
// make a specific error for when the migration lock is taken, change `(ms *Status) lock()` to return this when appropriate
var MigrationInProgress = fmt.Errorf("another process or system has the migration advisory lock 77")

if migrationNeeded {
  err = migrator.Start()
  if err != nil && err == migration. MigrationInProgress {
   // you might want to log the number you're using for the lock in case something gets stuck and a human wants to manually unstick it
    log.Info("another process or system is already running the migration, skipping it...")
  }
```

Reply:
Thank you for pointing that out. I was thinking that it would fail [here](https://github.com/chef/automate/pull/4219/files#diff-0d1121023e5ce205c1be3cc1b5ff83b9R74) if another service tried to start the migration at the same time. That would kill the service and restart it. I like your idea of recovering from the error though. I will update it.

Reply:
Oh, I thought the `pg_advisory_lock` would fail if it could not get the lock. Good call.

Reply:
@danielsdeleo I changed this to use the `pg_try_advisory_lock`command, but the service will restart if the try fails. I tested this out and I did not see a problem.